File: tools/rpc-clients/python/polycall_call.py
#!/usr/bin/env python3
"""A generic polycall_rpc v1 CLI: one round trip, no retry.

    python polycall_call.py <host:port> <service> <operation> <json-input> [deadline_ms]

Prints the RESPONSE payload exactly as received (for byte-identical
cross-language comparison in tests/conformance/) and exits with the same
code contract as the C CLI's `polycall call` (docs/RPC.md). Sibling of the
inventory-specific inventory_client.py (kept as the Stage 3 first-proof).
"""

import json
import socket
import struct
import sys

MAGIC = b"PCR1"
HEADER = struct.Struct(">4sBBHII")

EXIT_FOR = {
    "request.malformed": 3, "input.invalid": 3,
    "operation.unknown": 4, "item.unknown": 4,
    "deadline.exceeded": 6, "auth.denied": 7,
}


def recv_exact(sock: socket.socket, n: int) -> bytes:
    buf = b""
    while len(buf) < n:
        chunk = sock.recv(n - len(buf))
        if not chunk:
            raise ConnectionError("peer closed")
        buf += chunk
    return buf


def main() -> int:
    if len(sys.argv) < 4 or len(sys.argv) > 6:
        sys.stderr.write(
            "usage: polycall_call.py <host:port> <service> <operation> <json-input> [deadline_ms]\n")
        return 2
    endpoint, service, operation = sys.argv[1], sys.argv[2], sys.argv[3]
    raw_input = sys.argv[4] if len(sys.argv) >= 5 else "null"
    deadline_ms = int(sys.argv[5]) if len(sys.argv) == 6 else 5000

    try:
        input_value = json.loads(raw_input)
    except json.JSONDecodeError as e:
        sys.stderr.write(f"polycall-call: invalid JSON input: {e}\n")
        return 3

    host, _, port = endpoint.rpartition(":")
    host = host or "127.0.0.1"
    req = json.dumps(
        {"service": service, "operation": operation, "deadline_ms": deadline_ms,
         "input": input_value},
        separators=(",", ":"),
    ).encode("utf-8")

    try:
        with socket.create_connection((host, int(port)), timeout=(deadline_ms + 1500) / 1000) as sock:
            sock.sendall(HEADER.pack(MAGIC, 1, 0, 0, 1, len(req)) + req)
            magic, mtype, _f, _r, _c, length = HEADER.unpack(recv_exact(sock, HEADER.size))
            if magic != MAGIC:
                sys.stderr.write("polycall-call: bad frame magic\n")
                return 5
            payload = recv_exact(sock, length).decode("utf-8") if length else ""
    except socket.timeout:
        sys.stderr.write("polycall-call: no reply before the deadline\n")
        return 6
    except OSError as e:
        sys.stderr.write(f"polycall-call: transport: {e}\n")
        return 5

    if mtype != 2:
        sys.stderr.write(f"polycall-call: unexpected frame type {mtype}\n")
        return 5

    sys.stdout.write(payload + "\n")
    try:
        obj = json.loads(payload)
    except json.JSONDecodeError:
        return 1
    if obj.get("ok") is True:
        return 0
    code = (obj.get("error") or {}).get("code")
    return EXIT_FOR.get(code, 1)


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/rpc-clients/node/polycall_call.mjs
// A generic polycall_rpc v1 CLI: one round trip, no retry.
//
//   node polycall_call.mjs <host:port> <service> <operation> <json-input> [deadline_ms]
//
// Prints the RESPONSE payload exactly as received (for byte-identical
// cross-language comparison in tests/conformance/) and exits with the same
// code contract as the C CLI's `polycall call` (docs/RPC.md). Sibling of the
// inventory-specific inventory_client.mjs (kept as the Stage 3 first-proof).

import net from "node:net";

const [, , endpoint, service, operation, rawInput, rawDeadline] = process.argv;
if (!endpoint || !service || !operation) {
  process.stderr.write("usage: polycall_call.mjs <host:port> <service> <operation> <json-input> [deadline_ms]\n");
  process.exit(2);
}
const deadlineMs = rawDeadline ? Number(rawDeadline) : 5000;
let input;
try {
  input = JSON.parse(rawInput ?? "null");
} catch (e) {
  process.stderr.write(`polycall-call: invalid JSON input: ${e.message}\n`);
  process.exit(3);
}

const i = endpoint.lastIndexOf(":");
const host = endpoint.slice(0, i) || "127.0.0.1";
const port = Number(endpoint.slice(i + 1));

const EXIT = {
  "request.malformed": 3, "input.invalid": 3,
  "operation.unknown": 4, "item.unknown": 4,
  "deadline.exceeded": 6, "auth.denied": 7,
};

function frame(type, corr, payloadStr) {
  const payload = Buffer.from(payloadStr, "utf8");
  const hdr = Buffer.alloc(16);
  hdr.write("PCR1", 0, "ascii");
  hdr.writeUInt8(type, 4);
  hdr.writeUInt32BE(corr >>> 0, 8);
  hdr.writeUInt32BE(payload.length, 12);
  return Buffer.concat([hdr, payload]);
}

const reqPayload = JSON.stringify({ service, operation, deadline_ms: deadlineMs, input });
const sock = net.connect({ host, port }, () => sock.write(frame(1, 1, reqPayload)));
const timer = setTimeout(() => {
  process.stderr.write("polycall-call: no reply before the deadline\n");
  sock.destroy();
  process.exit(6);
}, deadlineMs + 1500);

let buf = Buffer.alloc(0);
sock.on("data", (d) => {
  buf = Buffer.concat([buf, d]);
  if (buf.length < 16) return;
  const len = buf.readUInt32BE(12);
  if (buf.length < 16 + len) return;
  clearTimeout(timer);
  const type = buf.readUInt8(4);
  const payload = buf.subarray(16, 16 + len).toString("utf8");
  sock.end();
  if (type !== 2) { process.stderr.write(`unexpected frame type ${type}\n`); process.exit(5); }
  process.stdout.write(payload + "\n");
  let ok = false, code = null;
  try { const o = JSON.parse(payload); ok = o.ok === true; code = o.error && o.error.code; } catch { /* ignore */ }
  process.exit(ok ? 0 : (EXIT[code] ?? 1));
});
sock.on("error", (e) => {
  clearTimeout(timer);
  process.stderr.write(`polycall-call: transport: ${e.message}\n`);
  process.exit(5);
});
